### DISCUSSION TITLE
Custom action button in relation manage form

### DISCUSSION THREAD

**Discussion by envas** [wintercms/winter#248] [Ideas]
Relation manage form has obviously hardcoded action buttons, either Create/Cancel or Update/Cancel, depending on the context. I didn't find any information about whether it is possible to add a custom button in the relation manage popup form via config. 

The only option that worked for me was to copy the partial _manage_form.htm from the behaviour to my controller, rename it to _relation_manage_form.htm and edit the form in my controller.

**Comment by bennothommo** [comment]
At this stage, modifying the partials of the Relation controller, or using a `partial` type field in the form config, are the only ways of adding custom buttons to the Relation manage form.

I'll convert this thread to an "Idea" thread, as it may be something we need to address in the future.

**Comment by envas** [comment]
Thank you for the feedback. If anyone is interested in this in the future, I have a note here.  

RelationController provides a lot of vars to the partial widget, but what I miss here is an unambiguous identification of the related model(s). When updating, the related model(s) can be addressed using relationship, because the `relationManageId` is available. I think currently it is impossible to find out related model(s) in the create operation.

**Reply by bennothommo** [reply]
@envas which model are you referring to? The related model being edited in the form, or the parent model that the relation model belongs to?

**Comment by envas** [comment]
The related model being edited in the popup form. 
I have a master-detail (Order - OrderItems) relation. On the Order form, I have a list of OrderItems. To Create/Update an item a popup with the OrderItem form fires. This is where I need to add my action button and pass the Order reference to my AJAX action function. On update, I can pass `relationManageId`, get the OrderItem model and find the matching Order. On create I did not find any reasonable reference to the parent model.

**Reply by bennothommo** [reply]
If it's within the Form context, I believe you can get the model. I think it's `$form->model` or `$this->formGetModel()` from memory.

**Comment by envas** [comment]
OK, got it. It was very simple in the end. Action method onMyButton receives the `recordId` of the parent. 

```
public function onMyButton($recordId) 
{
        $orderModel = $this->formFindModelObject($recordId);
       // do the rest
}
```

**Reply by LukeTowers** [reply]
I think this is because all controller AJAX methods receive the parameters provided to the current controller action. I.e. if your URL is `/author/plugin/controller/update/3` then `3` is the parameter provided to the `update($id)` method and it will also be provided to any AJAX methods called while within that action context. If you tried to call `onMyButton()` from the create context I believe you'd find that `$recordId` was not provided because it wouldn't exist as a URL parameter at that point.

**Reply by envas** [reply]
In my case, it works even for the create context. I  am creating a new child object (with the relation hasMany) in the parent form and my `onMyButton()` method is in the parent's controller. It's a bit crazy. When in a parent form with a detail list a new child entry is created/updated in a popup window, the button actions are still fired in the parent controller, not in the child controller.

**Reply by LukeTowers** [reply]
That sounds confusing, I'd have to actually see the code to understand what you're describing but if it works it works I guess 😄

**Reply by envas** [reply]
But that agrees with what you wrote earlier. When you edit the popup with the detail form, you still have the master id in the URL. And this is then passed as a parameter to the AJAX methods.
